### PR TITLE
Remove compile warnings under windows

### PR DESCRIPTION
**Pull request by bschommer**
This patch removes some compile warnings when compiling Ocaml with the Visual Studio under windows by adding:
- Define guards in the win32unix/unixsupport.c as in the other version
- Define guards in byterun/floats.c around isnan and isfinite

Also changed the definition of the Noreturn Macro in misc.h so that it also works with a Visual Studio version newer then 2008.

### CODE REVIEW

**Comment by gasche** [comment]
It seems there is an extraneous patch wrongly included in the PR (there was a small hiccup of the SVN/github mirror when this patch of Jeremy Yallop was merged, and I wonder whether this explains why I don't remember having seen your PR before). Would you rebase it against the current trunk and `git push --force` into your own branch to refresh the PR?

**Comment by bschommer** [comment]
I removed the extraneous patch.

**Comment by bschommer** [comment]
The potential compile warnings from the `CAMLparam` and `CAMLlocal` macros could also be removed when the initialization of the `caml__dummy_##x` variable

```
  CAMLunused int caml__dummy_##x = ( \
    (caml__roots_##x.next = caml_local_roots), \
    (caml_local_roots = &caml__roots_##x), \
    (caml__roots_##x.nitems = 1), \
    (caml__roots_##x.ntables = 1), \
    (caml__roots_##x.tables [0] = &x), \
    0)
```

is replaced by

```
  CAMLunused int caml__dummy_##x = ( \
    (caml__roots_##x.next = caml_local_roots), \
    (caml_local_roots = &caml__roots_##x), \
    (caml__roots_##x.nitems = 1), \
    (caml__roots_##x.ntables = 1), \
    (caml__roots_##x.tables [0] = &x), \
    ((void)caml__dummy_##x,
    0)
```

Which is valid in C. However I'm not sure if there are any C compiler that don't support this. I just tried it with gcc and clang and it seems okay.

File: otherlibs/win32unix/unixsupport.c [hunk 1]
```diff
@@ -170,45 +170,118 @@ void win32_maperr(DWORD errcode)
 }
 
 /* Windows socket errors */
-
+#ifndef EWOULDBLOCK
```
Comment (xavierleroy) [inline otherlibs/win32unix/unixsupport.c]: What happens if EWOULDBLOCK is already defined but is not equal to -WSAEWOULDBLOCK ?  Would that break the reporting of errors as a Unix_error exception?

Comment (bschommer) [inline otherlibs/win32unix/unixsupport.c]: You are right, erno defines them to different values. I think the better solution would be to just replace the uses of the macro in this file by the corresponding value.

**Comment by bschommer** [comment]
The last commit avoids the problem which occur when the error values are defined differently by just undefining them before redefining the values.

**Comment by damiendoligez** [comment]
> The define from trapsp to caml_trapsp in compatibility.h changed the trapsp to caml_trapsp and used the uninitialized variable to initialize itself.

Yes, I've seen this one (or another instance) recently (and I thought I'd fixed it) where `gcc` didn't warn, but MSVC complained. This makes me suspicious of your solution for `caml__dummy_##x` because it's basically the same thing: a use of the variable in its own initializer.

**Comment by bschommer** [comment]
I just tested the `caml__dummy##x` solution and it did not help. MSVC still complained about unused variables.

**Comment by bschommer** [comment]
I added some MSVC specific pragmas which disable the warning C4189 (unused local variable) for the `caml__dummy##x` variable.

File: byterun/caml/memory.h [hunk 2]
```diff
@@ -192,35 +192,48 @@ CAMLextern struct caml__roots_block *caml_local_roots;  /* defined in roots.c */
 
 
 #if defined(__GNUC__) && (__GNUC__ > 2 || (__GNUC__ == 2 && __GNUC_MINOR__ > 7))
-  #define CAMLunused __attribute__ ((unused))
+  #define CAMLunused_start __attribute__ ((unused))
```
Comment (damiendoligez) [inline byterun/caml/memory.h:157]: Why did you rename this macro?

**Comment by bschommer** [comment]
Since the pragma solution for windows requires a pusch before and a pop after I thought it would sense to rename the macro which adds the push into start and the one that ats the pop into end but I do not insist on this renaming.

**Comment by bschommer** [comment]
Since the pragma solution for windows requires a push before and a pop after I thought it would sense to rename the macro which adds the push into start and the one that ats the pop into end but I do not insist on this renaming.

**Comment by damiendoligez** [comment]
Oh that makes sense. I'd overlooked the "end" part, sorry.

**Comment by bschommer** [comment]
I updated the my pull request to get a working travis build again, the problems came from other commits.
Concerning the new names for the CAMLunused macro: Does somebody rely on them being in the header since otherwise I could add the old one back for compatibility reasons.

**Comment by gasche** [comment]
I merged the patch series in trunk (except for the trapsp one that is superseded by af15f03b42a0a95832bb734c1fd4df4c48dbb7c4 ). Thanks for the patches!

**Comment by alainfrisch** [comment]
This patch changes the way "Noreturn" should be used.  This breaks third-party stubs, for instance sqlite3-ocaml.  Perhaps we should think of a more backward compatible approach, keeping the old definition of Noreturn (ignored under MSVC) and providing a new macro with a different onem to be used in prefix position.

**Comment by gasche** [comment]
We could provide `Noreturn_start` and `Noreturn_end`, and have `CAMLunused` a compatibility alias of `CAMLunused_start` and `Noreturn` a compatibility alias of `Noreturn_end`. Isn't the `start/end` pair a bit heavy, though?

**Comment by bschommer** [comment]
You could also just add one Noreturn_internal or so for the header of the standard library and keep the old definition for the backward compatibility.

**Comment by gasche** [comment]
But it means that third-party stubs using the (non-internal) standard headers have no way to fix their own MSVC unused warnings.

**Comment by bschommer** [comment]
Yes you are right. But I would rename the new macro to `CAMLnoreturn` since C11 introduces a `noreturn` macro.

**Comment by gasche** [comment]
Would you propose a new patch(set) for whichever approach you think is best?

**Comment by bschommer** [comment]
No problem.

**Comment by gasche** [comment]
@alainfrisch `Noreturn` is now back to its previous semantics, restoring compatibility. Thanks for the report!
